### PR TITLE
Added Spiceblooms.

### PR DESCRIPTION
**Pull request by Unit158**
~~I currently need help rendering the spicebloom properly.~~ Closes https://github.com/OpenRA/OpenRA/issues/4462

### CODE REVIEW

**Comment by Unit158** [comment]
Note to reviewers: I have to yet to remove the code duplication.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+			explodedTicks = 0;
+			growing = true;
+			growTicks = 0;
+			respawnTime = self.World.SharedRandom.Next(info.RespawnTime[0], info.RespawnTime[1]);
+			growTime = self.World.SharedRandom.Next(info.GrowthTime[0], info.GrowthTime[1]);
+		}
+
+		void Explode()
+		{
+			//render.Remove(anim2);
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: > [SA1512] A single-line comment must not be followed by a blank line. To ignore this error when commenting out a line of code, begin the comment with '////' rather than '//'.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+					{
+						var projectile = args.Weapon.Projectile.Create(args);
+						if (projectile != null)
+							self.World.Add(projectile);
+
+							if (args.Weapon.Report != null && args.Weapon.Report.Any())
+								Sound.Play(args.Weapon.Report.Random(self.World.SharedRandom), self.CenterPosition);
+						}
+					});
+				}
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: > [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
+﻿using OpenRA.GameRules;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Support;
+using OpenRA.Traits;
+using System;
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: > [SA1208] System using directives must be placed before all other using directives.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
+﻿using OpenRA.GameRules;
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: Missing copyright header.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+				explodedTicks++;
+
+				if (explodedTicks >= respawnTime)
+					Respawn();
+			}
+		}
+
+		public void OnCrush(Actor crusher)
+		{
+			if(growing)
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: [SA1000] The spacing around the keyword 'if' is invalid.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+				Explode();
+		}
+
+		public void WarnCrush(Actor crusher) { }
+
+		public bool CrushableBy(string[] crushClasses, Player owner)
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: > [SP2000] Invalid spacing at the end of the line.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+			for (var i = 0; info.Pieces > i; i++)
+			{
+				var cell = RandomWalk(self.Location, random)
+					.Take(info.Range)
+					.SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p))
+					.Cast<CPos?>().RandomOrDefault(random);
+
+				if (cell == null)
+					cell = RandomWalk(self.Location, random).Take(info.Range).First();
+
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: > [SP2000] Invalid spacing at the end of the line.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+					.Cast<CPos?>().RandomOrDefault(random);
+
+				if (cell == null)
+					cell = RandomWalk(self.Location, random).Take(info.Range).First();
+				
+				var args = new ProjectileArgs
+				{
+					Weapon = wep,
+					Facing = 0,
+
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: > [SP2000] Invalid spacing at the end of the line.

File: mods/d2k/sequences/misc.yaml [hunk 9]
```diff
@@ -314,17 +314,19 @@ crate:
 
 # TODO: keep the redundant spicebloom.shp for now for the awful WinForms maps editor
```
Comment (Mailaender) [inline mods/d2k/sequences/misc.yaml]: We already axed it in https://github.com/OpenRA/OpenRA/pull/8511. I forgot to remove this line and `mods/d2k/bits/spicebloom.shp`. Would you be so kind and do it here?

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	public class SpicebloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
+	{
+		public readonly string[] GrowthSequences = { "grow1", "grow2", "grow3" };
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: Missing `[SequenceReference]`.

File: mods/d2k/rules/arrakis.yaml [hunk 11]
```diff
@@ -1,20 +1,19 @@
...
 	HiddenUnderShroud:
-	RenderBuilding:
-	Building:
-		Footprint: x
-		Dimensions: 1,1
+	RenderSprites:
 	AppearsOnRadar:
+		UseLocation: yes
+	EditorAppearance:
+		RelativeToTopLeft: yes
```
Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: This trait was removed in #8511 which is why it crashes the game.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+
+			if (resType == null)
+				throw new InvalidOperationException("No such resource type `{0}`".F(info.ResourceType));
+
+			resLayer = self.World.WorldActor.Trait<ResourceLayer>();
+
+			render = self.Trait<RenderSprites>();
+
+			anim = new Animation(init.Self.World, render.GetImage(self));
+			anim2 = new AnimationWithOffset(anim, null, () => false);
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: Instead of adding and removing it, you probably want to tie up the IsVisible func here.

File: OpenRA.Mods.D2k/Traits/Spicebloom.cs [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+	public class Spicebloom : ITick, ICrushable, INotifyDamage
+	{
+		readonly Actor self;
+		readonly SpicebloomInfo info;
+		readonly ResourceType resType;
+		readonly ResourceLayer resLayer;
+		readonly RenderSprites render;
+
+		Animation anim;
+		AnimationWithOffset anim2;
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Spicebloom.cs]: A more verbose name is preferred instead of enumerated variables names.

File: mods/d2k/rules/arrakis.yaml [hunk 14]
```diff
@@ -1,20 +1,19 @@
...
+	ProximityCaptor:
+		Types: Tree
```
Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: I don't think we add this back again. It was removed on purpose.

Comment (Unit158) [inline mods/d2k/rules/arrakis.yaml]: Thanks for the help, but I already knew about the style nits :/

Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: This is not a style problem. You chose the wrong version when resolving rebase conflicts.

Comment (Unit158) [inline mods/d2k/rules/arrakis.yaml]: That was supposed to be a general comment.

Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: The style problems may be spammy, but you completely ignored my remark that adding and removing animations is not what you want.

**Comment by Unit158** [comment]
Rendering now works correctly. I just need a bit of help tidying up the code now.

**Comment by Unit158** [comment]
This should now be in a reviewable state.

File: mods/d2k/rules/arrakis.yaml [hunk 15]
```diff
@@ -1,20 +1,25 @@
...
-	WithActiveAnimation:
+	Spicebloom:
+		Weapon: SpiceExplosion
 	RadarColorFromTerrain:
 		Terrain: Spice
-	BodyOrientation:
-	WithMakeAnimation:
+	Immobile:
+	Crushable:
+		CrushClasses: crate
```
Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: Use `spicebloom` here as well. You already added it to the units.

File: mods/d2k/rules/arrakis.yaml [hunk 16]
```diff
@@ -1,20 +1,25 @@
...
+	Health:
+		HP: 1
+		Radius: 512
+	TargetableUnit:
+		TargetTypes: Ground
+		RequiresForceFire: yes
```
Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: I guess the AI will have trouble here as it just attack moves through the map and needs auto target.

Comment (Mailaender) [inline mods/d2k/rules/arrakis.yaml]: Nevermind. I didn't notice this will explode by itself after a set amount of time.

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

File: OpenRA.Mods.Common/Traits/ThrowsShrapnel.cs [hunk 17]
```diff
@@ -18,9 +18,15 @@ namespace OpenRA.Mods.Common.Traits
 	public class ThrowsShrapnelInfo : ITraitInfo
 	{
 		[WeaponReference, FieldLoader.Require]
-		public string[] Weapons = { };
-		public int[] Pieces = { 3, 10 };
-		public WDist[] Range = { WDist.FromCells(2), WDist.FromCells(5) };
+		[Desc("The weapons used for shrapnel.")]
+		public readonly string[] Weapons = { };
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ThrowsShrapnel.cs]: This needs a `[WeaponReference]`.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/ThrowsShrapnel.cs]: Also, these documentation changes don't look directly related to spice blooms, so please split them into their own commit.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/ThrowsShrapnel.cs]: > This needs a `[WeaponReference]`.

It already has.

**Comment by Unit158** [comment]
Instead of spawning a husk, I'll just have the spicebloom remove itself before it grows again

**Comment by Unit158** [comment]
I'm no longer interested in developing this. Thanks to those who put their time in anyway.

**Comment by TheRaffy** [comment]
Makes me sad.

**Comment by Mailaender** [comment]
Superseded by https://github.com/OpenRA/OpenRA/pull/8987.
